From 689069d89b22dab480d0734ad55965e68be2d66e Mon Sep 17 00:00:00 2001
From: Aishwarya R <quic_aisr@quicinc.com>
Date: Thu, 2 Jan 2025 12:36:54 +0530
Subject: [PATCH] UPSTREAM:DP:CP:wifi: ath12k: Add support for dynamic DSCP to
 TID mapping at vdev level

Currently, ath12k follows fixed DSCP to TID mapping.Add support
for dynamic DSCP to TID mapping based on qos_map_set values from
mac80211

Get qos_map_set events from mac80211 and update TID registers
accordingly.

Parse the DSCP exception and UP(User Priority) values, DSCP low
and DSCP high range for every UP from mac80211. Update the TID
registers in vdev level based on the parsed values.

Below is the qos_map_set format
<DSCP Exceptions[DSCP,UP]>,]<UP 0 range[low,high]>,...<UP 7 range[low,high]

Patch-dependency: none
Signed-off-by: Aishwarya R <quic_aisr@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c   |   1 +
 drivers/net/wireless/ath/ath12k/core.h   |  28 +++++
 drivers/net/wireless/ath/ath12k/dp.c     |   2 +-
 drivers/net/wireless/ath/ath12k/hal_tx.c |  74 ++++++++++++
 drivers/net/wireless/ath/ath12k/hal_tx.h |   2 +
 drivers/net/wireless/ath/ath12k/mac.c    | 140 +++++++++++++++++++++++
 6 files changed, 246 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index b9f3e80..1f499f6 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -2319,6 +2319,7 @@ static void ath12k_core_pre_reconfigure_recovery(struct ath12k_base *ab)
 		}
 
 		ar->monitor_vdev_id = -1;
+		ar->free_map_id = ATH12K_FREE_MAP_ID_MASK;
 		clear_bit(MONITOR_VDEV_STARTED, &ar->monitor_flags);
 		clear_bit(MONITOR_VDEV_CREATED, &ar->monitor_flags);
 
diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 1f90fa1..8cb223b 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -125,6 +125,9 @@ extern bool ath12k_debug_critical;
 #define ATH12k_MLO_LINK_ID_INVALID 0xFF
 #define ATH12k_DS_NODE_ID_INVALID 0xFF
 
+#define ATH12K_MAX_TID_VALUE 8
+#define ATH12K_FREE_MAP_ID_MASK GENMASK(31, 0)
+
 enum ath12k_bdf_search {
     ATH12K_BDF_SEARCH_DEFAULT,
     ATH12K_BDF_SEARCH_BUS_AND_BOARD,
@@ -441,10 +444,15 @@ struct ath12k_recovery_cache {
 	u64 bss_conf_changed;
 };
 
+struct ath12k_cache_qos_map {
+	struct ath12k_qos_map *qos_map;
+};
+
 struct ath12k_vif_cache {
 	struct ath12k_tx_conf tx_conf;
 	struct ath12k_key_conf key_conf;
 	u64 bss_conf_changed;
+	struct ath12k_cache_qos_map cache_qos_map;
 };
 
 struct ath12k_vif_tcl_desc_template {
@@ -558,6 +566,9 @@ struct ath12k_link_vif {
 	bool is_link_removal_in_progress;
 	bool is_link_removal_update_pending;
 	struct ath12k_wmi_mlo_link_removal_event_params link_removal_data;
+	u8 map_id;
+	struct ath12k_qos_map *qos_map;
+	struct work_struct set_dscp_tid_work;
 };
 
 struct ath12k_vlan_iface {
@@ -567,6 +578,22 @@ struct ath12k_vlan_iface {
 	int ppe_vp_profile_idx[ATH12K_NUM_MAX_LINKS];
 };
 
+struct ath12k_dscp_range {
+	u8 low;
+	u8 high;
+};
+
+struct ath12k_dscp_exception {
+	u8 dscp;
+	u8 up;
+};
+
+struct ath12k_qos_map {
+	u8 num_des;
+	struct ath12k_dscp_exception dscp_exception[IEEE80211_QOS_MAP_MAX_EX];
+	struct ath12k_dscp_range up[ATH12K_MAX_TID_VALUE];
+};
+
 struct ath12k_vif {
 	enum wmi_vdev_type vdev_type;
 	struct ieee80211_vif *vif;
@@ -1294,6 +1321,7 @@ struct ath12k {
 	u32 scan_min_rest_time;
 	u32 scan_max_rest_time;
 	struct completion delete_all_peer_done;
+	u32 free_map_id;
 };
 
 struct ath12k_6ghz_sp_reg_rule {
diff --git a/drivers/net/wireless/ath/ath12k/dp.c b/drivers/net/wireless/ath/ath12k/dp.c
index edcb06e..58cb6d6 100644
--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -697,7 +697,7 @@ u32 ath12k_dp_tx_get_vdev_bank_config(struct ath12k_base *ab, struct ath12k_link
 			u32_encode_bits(vdev_id_check_en,
 					HAL_TX_BANK_CONFIG_VDEV_ID_CHECK_EN);
 
-	bank_config |= u32_encode_bits(0, HAL_TX_BANK_CONFIG_DSCP_TIP_MAP_ID);
+	bank_config |= u32_encode_bits(arvif->map_id, HAL_TX_BANK_CONFIG_DSCP_TIP_MAP_ID);
 
 	return bank_config;
 }
diff --git a/drivers/net/wireless/ath/ath12k/hal_tx.c b/drivers/net/wireless/ath/ath12k/hal_tx.c
index 467de34..fe5702a 100644
--- a/drivers/net/wireless/ath/ath12k/hal_tx.c
+++ b/drivers/net/wireless/ath/ath12k/hal_tx.c
@@ -10,6 +10,8 @@
 #include "hif.h"
 
 #define DSCP_TID_MAP_TBL_ENTRY_SIZE 64
+#define HAL_TX_BITS_PER_TID 3
+#define HAL_TX_NUM_DSCP_REG_SIZE 32
 
 /* dscp_tid_map - Default DSCP-TID mapping
  *=================
@@ -29,6 +31,69 @@ static inline u8 dscp2tid(u8 dscp)
 	return dscp >> 3;
 }
 
+void ath12k_update_dscp_register(struct ath12k_base *ab, u32 addr, u32 mask, u32 value)
+{
+	u32 reg_val;
+
+	reg_val = ath12k_hif_read32(ab, addr);
+	reg_val &= ~mask;
+	reg_val |= value;
+	ath12k_hif_write32(ab, addr, reg_val);
+}
+
+void ath12k_hal_tx_update_dscp_tid_map(struct ath12k_base *ab, int id, u8 dscp, u8 tid)
+{
+        u32 ctrl_reg_val;
+        u32 addr;
+        u32 start_index, end_index;
+        u32 mask;
+        u32 value;
+
+        ctrl_reg_val = ath12k_hif_read32(ab, HAL_SEQ_WCSS_UMAC_TCL_REG +
+                                         HAL_TCL1_RING_CMN_CTRL_REG);
+        /* Enable read/write access */
+        ctrl_reg_val |= HAL_TCL1_RING_CMN_CTRL_DSCP_TID_MAP_PROG_EN;
+        ath12k_hif_write32(ab, HAL_SEQ_WCSS_UMAC_TCL_REG +
+                           HAL_TCL1_RING_CMN_CTRL_REG, ctrl_reg_val);
+
+        addr = HAL_SEQ_WCSS_UMAC_TCL_REG + HAL_TCL1_RING_DSCP_TID_MAP +
+               (4 * id * (HAL_DSCP_TID_TBL_SIZE / 4));
+
+	/* Calculate start and end indices within the register */
+	start_index = dscp * HAL_TX_BITS_PER_TID;
+	end_index = (start_index + (HAL_TX_BITS_PER_TID - 1)) %
+                     HAL_TX_NUM_DSCP_REG_SIZE;
+	addr += (4 * (start_index / HAL_TX_NUM_DSCP_REG_SIZE));
+	start_index %= HAL_TX_NUM_DSCP_REG_SIZE;
+
+	if (end_index < start_index) {
+		/* Handle the case where the TID value spans two registers */
+		mask = GENMASK((HAL_TX_NUM_DSCP_REG_SIZE - 1), start_index);
+		value = (tid << start_index) & mask;
+		/* Update the first register */
+		ath12k_update_dscp_register(ab, addr, mask, value);
+
+		/* Update the second register */
+		addr = addr+4;
+		mask = GENMASK(end_index, 0);
+		value = (tid >> (HAL_TX_NUM_DSCP_REG_SIZE - start_index)) & mask;
+		ath12k_update_dscp_register(ab, addr, mask, value);
+	} else {
+		/* Handle the case where the TID value fits within one register */
+		mask = GENMASK(end_index, start_index);
+		value = (tid << start_index) & mask;
+		ath12k_update_dscp_register(ab, addr, mask, value);
+	}
+
+	/* Disable read/write access */
+	ctrl_reg_val = ath12k_hif_read32(ab, HAL_SEQ_WCSS_UMAC_TCL_REG +
+					 HAL_TCL1_RING_CMN_CTRL_REG);
+	ctrl_reg_val &= ~HAL_TCL1_RING_CMN_CTRL_DSCP_TID_MAP_PROG_EN;
+	ath12k_hif_write32(ab, HAL_SEQ_WCSS_UMAC_TCL_REG +
+			   HAL_TCL1_RING_CMN_CTRL_REG,
+			   ctrl_reg_val);
+} 
+
 void ath12k_hal_tx_set_dscp_tid_map(struct ath12k_base *ab, int id)
 {
 	u32 ctrl_reg_val;
@@ -107,6 +172,15 @@ void ath12k_hal_tx_configure_bank_register(struct ath12k_base *ab, u32 bank_conf
 			   bank_config);
 }
 
+u32 ath12k_hal_tx_read_bank_register(struct ath12k_base *ab, u8 bank_id)
+{
+	u32 reg_val;
+
+	reg_val = ath12k_hif_read32(ab, HAL_TCL_SW_CONFIG_BANK_ADDR + 4 * bank_id);
+
+	return reg_val;
+}
+
 void ath12k_hal_tx_vdev_mcast_ctrl_set(struct ath12k_base *ab, u32 vdev_id,
 				       u8 mcast_ctrl_val)
 {
diff --git a/drivers/net/wireless/ath/ath12k/hal_tx.h b/drivers/net/wireless/ath/ath12k/hal_tx.h
index 057b570..96af5c5 100644
--- a/drivers/net/wireless/ath/ath12k/hal_tx.h
+++ b/drivers/net/wireless/ath/ath12k/hal_tx.h
@@ -215,10 +215,12 @@ struct hal_tx_fes_status_end {
 void ath12k_hal_tx_set_ppe_vp_entry(struct ath12k_base *ab, u32 ppe_vp_config,
                                     u32 ppe_vp_index);
 #endif
+void ath12k_hal_tx_update_dscp_tid_map(struct ath12k_base *ab, int id, u8 dscp, u8 tid);
 void ath12k_hal_tx_set_dscp_tid_map(struct ath12k_base *ab, int id);
 int ath12k_hal_reo_cmd_send(struct ath12k_base *ab, struct hal_srng *srng,
 			    enum hal_reo_cmd_type type,
 			    struct ath12k_hal_reo_cmd *cmd);
+u32 ath12k_hal_tx_read_bank_register(struct ath12k_base *ab, u8 bank_id);
 void ath12k_hal_tx_configure_bank_register(struct ath12k_base *ab, u32 bank_config,
 					   u8 bank_id);
 void ath12k_hal_tx_vdev_mcast_ctrl_set(struct ath12k_base *ab, u32 vdev_id,
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index a57181d..be018cb 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -67,6 +67,7 @@ static u8 ath12k_mac_ahsta_get_pri_link_id(struct ath12k_vif *ahvif,
 					   struct ath12k_sta *ahsta,
 					   unsigned long int valid_links);
 static void ath12k_wmi_migration_cmd_work(struct work_struct *work);
+static void ath12k_set_dscp_tid_work(struct work_struct *wk);
 
 static const struct ieee80211_channel ath12k_2ghz_channels[] = {
 	CHAN2G(1, 2412, 0),
@@ -5853,6 +5854,7 @@ vdev_clean_up:
 	ath12k_dbg(ar->ab, ATH12K_DBG_SET(MAC, L1), "mac vdev delete id %d type %d subtype %d map %llx\n",
 		   arvif->vdev_id, ahvif->vdev_type, arvif->vdev_subtype,
 		   ar->ab->free_vdev_map);
+	ar->free_map_id |= 1 << arvif->map_id;
 
 clean_up:
 	arvif->is_created = false;
@@ -6074,6 +6076,8 @@ ath12k_mac_assign_link_vif(struct ath12k_hw *ah, struct ieee80211_vif *vif,
 			init_completion(&arvif->wmi_migration_event_resp);
 			INIT_WORK(&arvif->wmi_migration_cmd_work,
 				  ath12k_wmi_migration_cmd_work);
+			INIT_WORK(&arvif->set_dscp_tid_work,
+				  ath12k_set_dscp_tid_work);
 			INIT_LIST_HEAD(&arvif->peer_migrate_list);
 
 			if (!is_bridge_vdev) {
@@ -9957,6 +9961,124 @@ static void ath12k_mac_op_sta_set_4addr(struct ieee80211_hw *hw,
 	}
 }
 
+static void ath12k_mac_op_set_dscp_tid(struct ieee80211_hw *hw,
+				       struct ieee80211_vif *vif,
+				       struct cfg80211_qos_map *qos_map,
+				       unsigned int link_id)
+{
+	struct ath12k_hw *ah = hw->priv;
+	struct ath12k_vif *ahvif = ath12k_vif_to_ahvif(vif);
+	struct ath12k_link_vif *arvif;
+	struct ath12k_qos_map *new_qos_map;
+	struct ath12k *ar;
+
+	mutex_lock(&ah->conf_mutex);
+	new_qos_map = kzalloc(sizeof(*new_qos_map), GFP_KERNEL);
+	if (!new_qos_map) {
+		mutex_unlock(&ah->conf_mutex);
+		return;
+	}
+	memcpy(new_qos_map, qos_map, sizeof(*qos_map));
+	ahvif->cache[link_id].cache_qos_map.qos_map = NULL;
+
+	arvif = ath12k_get_arvif_from_link_id(ahvif, link_id);
+	if (arvif == NULL || !arvif->is_created) {
+		ath12k_info(NULL,
+			    "qos map changes cached to apply after vdev create\n");
+		ahvif->cache[link_id].cache_qos_map.qos_map = new_qos_map;
+		mutex_unlock(&ah->conf_mutex);
+		return;
+	}
+
+	ar = arvif->ar;
+	if (!ar) {
+		ath12k_err(NULL, "Failed to set DSCP to TID mapping\n");
+		mutex_unlock(&ah->conf_mutex);
+		kfree(new_qos_map);
+		return;
+	}
+
+	spin_lock_bh(&ar->data_lock);
+	arvif->qos_map = new_qos_map;
+	ieee80211_queue_work(ah->hw, &arvif->set_dscp_tid_work);
+	spin_unlock_bh(&ar->data_lock);
+	mutex_unlock(&ah->conf_mutex);
+}
+
+static void ath12k_mac_update_qos_map(struct ath12k *ar, struct ath12k_link_vif *arvif)
+{
+	struct ath12k_qos_map *qos_map;
+	u8 dscp_low, dscp_high;
+	u8 dscp;
+	u8 tid, map_id, bank_id;
+	u8 i;
+
+	qos_map = arvif->qos_map;
+	map_id = arvif->map_id;
+	bank_id = arvif->bank_id;
+
+	if (map_id >= HAL_DSCP_TID_MAP_TBL_NUM_ENTRIES_MAX) {
+		ath12k_err(ar->ab, "failed to find free map_id\n");
+		goto free_qos_map;
+	}
+
+	if (bank_id == DP_INVALID_BANK_ID) {
+		ath12k_err(ar->ab, "unable to find TX bank profile\n");
+		goto free_qos_map;
+	}
+
+	for (i = 0; i < ATH12K_MAX_TID_VALUE; i++) {
+		dscp_low = qos_map->up[i].low;
+		dscp_high = qos_map->up[i].high;
+		tid = i;
+
+		ath12k_dbg(ar->ab, ATH12K_DBG_MAC, "dscp_low:%d, dscp_high:%d, tid:%d, map_id:%d, bank_id:%d\n",
+			   dscp_low, dscp_high, tid, map_id, bank_id);
+		if (dscp_low == 0xFF || dscp_high == 0xFF)
+			continue;
+		for (dscp = dscp_low; dscp <= dscp_high; dscp++) {
+			ath12k_hal_tx_update_dscp_tid_map(ar->ab, map_id, dscp, tid);
+		}
+	}
+
+	if (qos_map->num_des > 0) {
+		for (i = 0; i < qos_map->num_des; i++) {
+			dscp = qos_map->dscp_exception[i].dscp;
+			tid = qos_map->dscp_exception[i].up;
+
+			ath12k_dbg(ar->ab, ATH12K_DBG_MAC, "dscp:%d, tid: %d, map_id:%d, bank_id:%d\n",
+				   dscp, tid, map_id, bank_id);
+			if (dscp == 0xFF)
+				continue;
+			ath12k_hal_tx_update_dscp_tid_map(ar->ab, map_id, dscp, tid);
+		}
+	}
+free_qos_map:
+	kfree(qos_map);
+	qos_map = NULL;
+	return;
+}
+
+static void ath12k_set_dscp_tid_work(struct work_struct *wk)
+{
+	struct ath12k_link_vif *arvif = container_of(wk, struct ath12k_link_vif, set_dscp_tid_work);
+	struct ath12k_qos_map *qos_map;
+	struct ath12k *ar;
+
+	qos_map = arvif->qos_map;
+	ar = arvif->ar;
+	if (!ar) {
+		ath12k_err(NULL, "Failed to set DSCP to TID mapping\n");
+		kfree(qos_map);
+		qos_map = NULL;
+		return;
+	}
+
+	spin_lock_bh(&ar->data_lock);
+	ath12k_mac_update_qos_map(ar, arvif);
+	spin_unlock_bh(&ar->data_lock);
+}
+
 static void ath12k_mac_op_sta_rc_update(struct ieee80211_hw *hw,
 					struct ieee80211_vif *vif,
 					struct ieee80211_sta *sta,
@@ -12986,6 +13108,7 @@ int ath12k_mac_vdev_create(struct ath12k *ar, struct ath12k_link_vif *arvif,
 	u8 link_addr[ETH_ALEN];
 	int txpower = NL80211_TX_POWER_AUTOMATIC;
 	u8 link_id;
+	u8 map_id;
 
 	lockdep_assert_held(&ar->conf_mutex);
 
@@ -13070,6 +13193,9 @@ int ath12k_mac_vdev_create(struct ath12k *ar, struct ath12k_link_vif *arvif,
 	arvif->dp = &ab->dp;
 	arvif->dev = ab->dev;
 	arvif->pdev_idx = ar->pdev_idx;
+	map_id =  __ffs(ar->free_map_id);
+	ar->free_map_id &= ~(1 << map_id);
+	arvif->map_id = map_id;
 
 	switch (vif->type) {
 	case NL80211_IFTYPE_UNSPECIFIED:
@@ -13116,6 +13242,7 @@ int ath12k_mac_vdev_create(struct ath12k *ar, struct ath12k_link_vif *arvif,
 	        spin_lock_bh(&ar->ab->base_lock);
 		ab->free_vdev_map |= 1LL << arvif->vdev_id;
 		spin_unlock_bh(&ar->ab->base_lock);
+		ar->free_map_id |= 1 << arvif->map_id;
 
 		goto err;
 	}
@@ -13290,6 +13417,7 @@ err_vdev_del:
 		ar->num_created_vdevs--;
 	arvif->is_created = false;
 	ar->allocated_vdev_map &= ~(1LL << arvif->vdev_id);
+	ar->free_map_id |= 1 << arvif->map_id;
 	spin_lock_bh(&ar->ab->base_lock);
 	ab->free_vdev_map |= 1LL << arvif->vdev_id;
 	spin_unlock_bh(&ar->ab->base_lock);
@@ -13346,6 +13474,13 @@ void ath12k_mac_vif_cache_flush(struct ath12k *ar,  struct ieee80211_vif *vif,
 		cache->bss_conf_changed = 0;
 	}
 
+	if (cache->cache_qos_map.qos_map) {
+		spin_lock_bh(&ar->data_lock);
+		arvif->qos_map = cache->cache_qos_map.qos_map;
+		ath12k_mac_update_qos_map(ar, arvif);
+		spin_unlock_bh(&ar->data_lock);
+	}
+
 	if (!list_empty(&cache->key_conf.list)) {
 		list_for_each_entry_safe(key_conf, tmp, &cache->key_conf.list, list) {
 			if (key_conf->sta) {
@@ -13672,6 +13807,7 @@ ppe_vp_config:
 	init_completion(&ahvif->deflink.wmi_migration_event_resp);
 	INIT_WORK(&ahvif->deflink.wmi_migration_cmd_work, ath12k_wmi_migration_cmd_work);
 	INIT_LIST_HEAD(&ahvif->deflink.peer_migrate_list);
+	INIT_WORK(&ahvif->deflink.set_dscp_tid_work, ath12k_set_dscp_tid_work);
 
 	ahvif->key_cipher = INVALID_CIPHER;
 
@@ -13735,6 +13871,7 @@ static void ath12k_mac_remove_link_interface(struct ieee80211_hw *hw,
 	}
 	cancel_work_sync(&arvif->peer_ch_width_switch_work);
 	cancel_work_sync(&arvif->wmi_migration_cmd_work);
+	cancel_work_sync(&arvif->set_dscp_tid_work);
 	cancel_delayed_work_sync(&ahvif->deflink.connection_loss_work);
 
 	lockdep_assert_held(&ah->conf_mutex);
@@ -19756,6 +19893,7 @@ static const struct ieee80211_ops ath12k_ops = {
 	.set_key                        = ath12k_mac_op_set_key,
 	.sta_state                      = ath12k_mac_op_sta_state,
 	.sta_set_4addr                  = ath12k_mac_op_sta_set_4addr,
+	.set_dscp_tid			= ath12k_mac_op_set_dscp_tid,
 	.sta_set_txpwr			= ath12k_mac_op_sta_set_txpwr,
 	.sta_rc_update			= ath12k_mac_op_sta_rc_update,
 	.conf_tx                        = ath12k_mac_op_conf_tx,
@@ -20495,6 +20633,7 @@ static int __ath12k_mac_register(struct ath12k *ar)
 
 	ar->max_num_stations = TARGET_NUM_STATIONS(ar->ab);
 	ar->max_num_peers = TARGET_NUM_PEERS_PDEV(ar->ab);
+	ar->free_map_id = ATH12K_FREE_MAP_ID_MASK;
 
 	total_vdevs = ath12k_core_get_total_num_vdevs(ar->ab);
 	if (total_vdevs == ATH12K_MAX_NUM_VDEVS_NLINK)
@@ -20836,6 +20975,7 @@ static int ath12k_mac_hw_register(struct ath12k_hw *ah)
 	ieee80211_hw_set(hw, SUPPORTS_TID_CLASS_OFFLOAD);
 	ieee80211_hw_set(hw, SUPPORTS_MULTI_BSSID);
 	ieee80211_hw_set(hw, HAS_TX_QUEUE);
+	ieee80211_hw_set(hw, SUPPORTS_DSCP_TID_MAP);
 
 	if (ath12k_frame_mode == ATH12K_HW_TXRX_ETHERNET) {
 		ieee80211_hw_set(hw, SUPPORTS_TX_ENCAP_OFFLOAD);
-- 
2.34.1

